### PR TITLE
Refactor message processing architecture

### PR DESCRIPTION
**Pull request by davidkrauser**
## Summary

This PR introduces a significant architectural enhancement that separates message moderation into two distinct phases: **asynchronous content analysis** and **post-processing actions**. This decoupling improves performance, respects API rate limits, and enables better caching and error handling.

**This change is necessary for future planned changes to block notifications with harmful content.** Apologies for the large size of this change-set - this is a very large architectural change that couldn't really be broken up.

I used Claude Code to help generate this PR description. Please excuse the excessive use of emojis :sweat_smile: 

### 🔄 **Dual-Processor Architecture**

**Previous Architecture:**
- Single `processor.go` handled both moderation analysis and post actions synchronously
- All processing occurred in `MessageHasBeenPosted`/`MessageHasBeenUpdated` hooks

**New Architecture:**
- **`ModerationProcessor`** (`server/message_processor.go`): Asynchronously analyzes message content using Azure AI Content Safety API
- **`PostProcessor`** (`server/post_processor.go`): Handles post-moderation actions (deletion, notifications, audit logging)
- **`ModerationResultsCache`** (`server/moderation_results_cache.go`): Caches moderation results and coordinates between processors

### 🎯 **Hook Strategy Changes**

The new architecture leverages a comprehensive hook strategy to implement asynchronous processing:

- **`MessageWillBePosted`** and **`MessageWillBeUpdated`**: These hooks now serve as the entry point for content analysis. When a message is about to be posted or updated, we immediately queue the message text for asynchronous moderation analysis. This allows the message to be posted without blocking while analysis happens in the background.

- **`MessageHasBeenPosted`** and **`MessageHasBeenUpdated`**: These hooks handle post-processing actions. They queue the complete post object for processing, which then waits for the moderation analysis result (initiated by the "Will" hooks) before taking any actions like deletion or user notification.

This approach ensures that message posting remains responsive while still providing comprehensive content moderation through the coordination between the async moderation analysis and the post-processing workflow.

### 📊 **Rate Limiting & Performance**

- **Azure API Rate Limiting**: Respects 1000 requests/minute limit by processing at 500 requests/minute
- **Queue Management**: 10,000 message queue size with overflow protection
- **Cache TTL**: 5-minute cache expiration for moderation results
- **Timeout Handling**: 15-second API timeout with 1-minute post-processing timeout

### 🧠 **ModerationProcessor** (`server/message_processor.go`)
```go
type ModerationProcessor struct {
    moderator              moderation.Moderator
    thresholdValue         int
    moderationResultsCache *moderationResultsCache
    messagesCh             chan string          // Rate-limited message queue
    done                   chan struct{}        // Graceful shutdown
    cleanupTicker          *time.Ticker        // Cache cleanup
}
```

**Responsibilities:**
- Queue messages for async analysis (`queueMessage`)
- Rate-limited API calls to Azure Content Safety
- Cache moderation results for post-processor consumption
- Automatic cache cleanup every 5 minutes

### 💾 **ModerationResultsCache** (`server/moderation_results_cache.go`)
```go
type moderationResultsCache struct {
    cache     map[string]*moderationResult     // Message -> Result mapping
    listeners map[string][]chan *moderationResult // Result notification system
    cacheTTL  time.Duration                    // 5-minute expiration
    cacheLock sync.Mutex                       // Thread-safe operations
}
```

**Features:**
- **Deduplication**: Prevents duplicate analysis of identical messages
- **Notification System**: `waitForResult()` allows post-processor to wait for analysis completion
- **Result States**: `pending`, `processed`, `flagged`, `error`
- **Automatic Cleanup**: Expired entries removed every 5 minutes

### 📮 **PostProcessor** (`server/post_processor.go`)
- **Enhanced from previous `processor.go`**
- Now waits for moderation results instead of performing analysis
- Maintains all existing functionality: post deletion, user notifications, audit logging
- Uses `waitForResult(message, timeout)` to coordinate with ModerationProcessor

### ✅ **New Test Files**
- **`server/configuration_test.go`**: Configuration parsing and validation (175 lines)
- **`server/message_processor_test.go`**: ModerationProcessor logic and threshold handling (133 lines)  
- **`server/moderation_results_cache_test.go`**: Cache operations, notifications, cleanup (251 lines)
- **`server/post_processor_test.go`**: Post-processing workflows

### 🔍 **Test Coverage Areas**
- Configuration parsing (`ExcludedUserSet`, `ExcludedChannelSet`, `ThresholdValue`)
- Threshold comparison logic with edge cases
- Cache lifecycle (pending → processed/flagged/error)
- Notification system with multiple waiters
- Cache cleanup and expiration
- Rate limiting and queue overflow handling

### 🚀 **Performance Improvements**
- **Non-blocking Message Flow**: Messages post immediately while analysis happens asynchronously
- **Efficient Caching**: Identical messages analyzed only once
- **Rate Limit Compliance**: Prevents API throttling and service degradation

### 🛡️ **Reliability Enhancements**  
- **Graceful Error Handling**: API failures don't block message posting
- **Resource Management**: Bounded queues prevent memory issues
- **Cleanup Automation**: Prevents cache bloat with automatic expiration

### 🔧 **Maintainability**
- **Separation of Concerns**: Clear boundaries between analysis and action phases
- **Comprehensive Testing**: 85%+ test coverage on new components
- **Extensible Design**: Easy to add new moderation providers or actions

### CODE REVIEW

**Review by BenCookie95 — APPROVED** [review]
Nice work, just a minor comment

**Review by fmartingr — COMMENTED** [review]
LGTM. Can you add some documentation for this in the README or a separate file with a small mermaid diagram or how data flows? I could follow this today but you never know six months from now 😅 I think you could easily ask the LLM to generate the documentation for you like you did with the PR description.

**Review by fmartingr — APPROVED** [review]
Forgot to click the appropriate radio button.

File: server/hooks.go [hunk 1]
```diff
@@ -5,14 +5,28 @@ import (
 	"github.com/mattermost/mattermost/server/public/plugin"
 )
 
+func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
+	if p.moderationProcessor != nil {
+		p.moderationProcessor.queueMessage(p.API, post.Message)
+	}
+	return nil, ""
```
[the remark this answers is not in the document]
Comment (davidkrauser) [reply server/hooks.go:12]: From the docs: `To allow the post without modification, return a nil *model.Post and an empty string.`

File: server/plugin.go [hunk 2]
```diff
@@ -71,28 +72,36 @@ func (p *Plugin) initialize(config *configuration) error {
...
 
 	processor, err := newPostProcessor(
-		botID, moderator, config.AuditLoggingEnabled,
-		thresholdValue, excludedUsers, excludedChannels,
+		botID, config.AuditLoggingEnabled, moderationResultsCache,
+		excludedUsers, excludedChannels,
 		config.ExcludeDirectMessages, config.ExcludePrivateChannels)
 	if err != nil {
-		return errors.Wrap(err, "failed to create post processor")
+		return errors.Wrap(err, "failed to create post post processor")
```
[the remark this answers is not in the document]
Comment (davidkrauser) [reply server/plugin.go:101]: done :+1:

File: server/message_processor.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/mattermost/mattermost/server/public/plugin"
+)
+
+// The current Azure rate limit is 1000 posts per minute.
+// Using half of that to give us some wiggle room:
+// https://learn.microsoft.com/en-us/azure/ai-services/content-safety/faq
+const (
+	moderationAPITimeout             = 15 * time.Second
+	maxModerationProcessingQueueSize = 10000
+	moderationsPerMinuteLimit        = 500
```
[the remark this answers is not in the document]
Comment (davidkrauser) [reply server/message_processor.go:17]: That's a good idea. Made a task here: https://github.com/mattermost/mattermost-plugin-content-moderation/issues/24
